api/v1alpha1/clusterexternaldns_types.go

```diff
@@ -0,0 +1,92 @@
+package v1alpha1
+
+import (
+	"github.com/Azure/aks-app-routing-operator/api"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterExternalDNS allows users to specify desired the state of a cluster-scoped ExternalDNS deployment and includes information about the state of their resources in the form of Kubernetes events.
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,shortName=cedns
+type ClusterExternalDNS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterExternalDNSSpec   `json:"spec,omitempty"`
+	Status ClusterExternalDNSStatus `json:"status,omitempty"`
+}
+
+func (c *ClusterExternalDNS) GetCondition(conditionType string) *metav1.Condition {
+	return meta.FindStatusCondition(c.Status.Conditions, conditionType)
+}
+
+func (c *ClusterExternalDNS) GetConditions() *[]metav1.Condition {
+	return &c.Status.Conditions
+}
+
+func (c *ClusterExternalDNS) GetGeneration() int64 {
+	return c.Generation
+}
+
+func (c *ClusterExternalDNS) SetCondition(condition metav1.Condition) {
+	api.VerifyAndSetCondition(c, condition)
+}
+
+// ClusterExternalDNSSpec allows users to specify desired the state of a cluster-scoped ExternalDNS deployment.
+type ClusterExternalDNSSpec struct {
+	// TenantID is the ID of the Azure tenant where the DNS zones are located.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Format:=uuid
+	// +kubebuilder:validation:Pattern=`[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`
+	TenantID string `json:"tenantID"`
+
+	// DNSZoneResourceIDs is a list of Azure Resource IDs of the DNS zones that the ExternalDNS controller should manage. These must be in the same resource group and be of the same type (public or private).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:MaxItems:=20
+	// +kubebuilder:validation:items:Pattern:=`(?i)\/subscriptions\/(.{36})\/resourcegroups\/(.+?)\/providers\/Microsoft.network\/(dnszones|privatednszones)\/(.+)`
```

Comment:
can we validate that all the resource groups / subs are the same with cel. might be possible

Reply:
I was thinking it'd be easier/more intuitive to have that in the controller + emit events if not

Reply:
it's better at this level is possible because it blocks the resource from being even created and why. Eventing is an okay mechanism for reporting information but for invalid specs it's much much better to block at the create level